phase4-entsog-client/src/main/java/com/helger/phase4/entsog/Phase4ENTSOGSender.java

```diff
@@ -0,0 +1,191 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.OverridingMethodsMustInvokeSuper;
+import javax.annotation.concurrent.Immutable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.helger.phase4.attachment.Phase4OutgoingAttachment;
+import com.helger.phase4.attachment.WSS4JAttachment;
+import com.helger.phase4.client.AS4ClientUserMessage;
+import com.helger.phase4.crypto.ECryptoKeyIdentifierType;
+import com.helger.phase4.sender.AS4BidirectionalClientHelper;
+import com.helger.phase4.sender.AbstractAS4UserMessageBuilder;
+import com.helger.phase4.util.AS4ResourceHelper;
+import com.helger.phase4.util.Phase4Exception;
+
+/**
+ * This class contains all the specifics to send AS4 messages with the ENTSOG
+ * profile. See <code>sendAS4Message</code> as the main method to trigger the
+ * sending, with all potential customization.
+ *
+ * @author Pavel Rotek
+ * @since 0.13.3
+ */
+@Immutable
+public final class Phase4ENTSOGSender {
+	private static final Logger LOGGER = LoggerFactory.getLogger(Phase4ENTSOGSender.class);
+
+	private Phase4ENTSOGSender() {
+	}
+
+	/**
+	 * @return Create a new Builder for AS4 messages if the payload is present.
+	 *         Never <code>null</code>.
+	 */
+	@Nonnull
+	public static ENTSOGUserMessageBuilder builder() {
+		return new ENTSOGUserMessageBuilder();
+	}
+
+	/**
+	 * Abstract ENTSOG UserMessage builder class with sanity methods
+	 *
+	 * @author Philip Helger
+	 * @param <IMPLTYPE> The implementation type
+	 */
+	public abstract static class AbstractENTSOGUserMessageBuilder<IMPLTYPE extends AbstractENTSOGUserMessageBuilder<IMPLTYPE>>
+			extends AbstractAS4UserMessageBuilder<IMPLTYPE> {
+
+		private ECryptoKeyIdentifierType m_signingKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		private ECryptoKeyIdentifierType m_encryptionKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		
+		private Phase4OutgoingAttachment m_aPayload;
+		private Charset m_aPayloadCharset;
+		private String m_aPayloadDocumentType;
+		
+		protected AbstractENTSOGUserMessageBuilder() {
+			// Override default values
+			try {
+				httpClientFactory(new Phase4ENTSOGHttpClientSettings());
+			} catch (final Exception ex) {
+				throw new IllegalStateException("Failed to init AS4 Client builder", ex);
+			}
+		}
+
+		/**
+		 * Encryption Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setEncryptionKeyIdentifierType(ECryptoKeyIdentifierType encryptionKeyIdentifierType) {
+			m_encryptionKeyIdentifierType = encryptionKeyIdentifierType;
+			return thisAsT();
+		}
+
+		/**
+		 * Signing Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setSigningKeyIdentifierType(ECryptoKeyIdentifierType signingKeyIdentifierType) {
```

Comment:
And here eventually the `AS4SigningParams`?

Reply:
I gave it a second thought and I think it should stay as it is. Because if we take it up the hierarchy, this leads to the problem that users of other profiles might accidentally use it. If this is the only parameter that should be changeable - so be it.